api/layer/multipart_upload.go

```diff
@@ -0,0 +1,563 @@
+			Reader: pr,
+			Header: initMetadata,
+		})
+		if err != nil {
+			n.log.Error("could not put a completed object (multipart upload)",
+				zap.String("uploadID", p.Info.UploadID),
+				zap.String("uploadKey", p.Info.Key),
+				zap.Error(err))
+			done <- true
+			return
+		}
+		uploadCompleted = true
+		done <- true
+	}(done)
+	for _, part := range parts {
+		_, err := n.objectGetWithPayloadWriter(ctx, &getParams{
+			Writer: pw,
+			cid:    p.Info.Bkt.CID,
+			oid:    part.ID,
+		})
+		if err != nil {
+			_ = pw.Close()
+			n.log.Error("could not download a part of multipart upload",
+				zap.String("uploadID", p.Info.UploadID),
+				zap.String("part number", part.Headers[UploadPartNumberAttributeName]),
+				zap.Error(err))
+			return nil, err
+		}
+	}
+	_ = pw.Close()
+	<-done
+
+	if !uploadCompleted {
+		return nil, errors.GetAPIError(errors.ErrInternalError)
+	}
+
+	for partNum, objInfo := range objects {
+		if partNum == 0 {
+			continue
+		}
+		if err = n.objectDelete(ctx, p.Info.Bkt.CID, objInfo.ID); err != nil {
+			n.log.Warn("could not delete upload part",
+				zap.Stringer("object id", objInfo.ID),
+				zap.Stringer("bucket id", p.Info.Bkt.CID),
+				zap.Error(err))
+			return nil, errors.GetAPIError(errors.ErrInternalError)
+		}
+	}
+
+	return obj, nil
+}
+
+func (n *layer) ListMultipartUploads(ctx context.Context, p *ListMultipartUploadsParams) (*ListMultipartUploadsInfo, error) {
+	var result ListMultipartUploadsInfo
+	if p.MaxUploads == 0 {
+		return &result, nil
+	}
+
+	f := &findParams{
+		filters: []filter{{attr: UploadPartNumberAttributeName, val: "0"}},
+		cid:     p.Bkt.CID,
+	}
+
+	ids, err := n.objectSearch(ctx, f)
+	if err != nil {
+		return nil, err
+	}
+
+	uploads := make([]*UploadInfo, 0, len(ids))
+
+	for _, id := range ids {
+		meta, err := n.objectHead(ctx, p.Bkt.CID, id)
+		if err != nil {
+			n.log.Warn("couldn't head object",
+				zap.Stringer("object id", id),
+				zap.Stringer("bucket id", p.Bkt.CID),
+				zap.Error(err))
+			continue
+		}
+		info := uploadInfoFromMeta(meta, p.Prefix, p.Delimiter)
+		if info != nil {
+			uploads = append(uploads, info)
+		}
+	}
+
+	sort.Slice(uploads, func(i, j int) bool {
+		if uploads[i].Key == uploads[j].Key {
+			return uploads[i].UploadID < uploads[j].UploadID
+		}
+		return uploads[i].Key < uploads[j].Key
+	})
+
+	if p.KeyMarker != "" {
+		if p.UploadIDMarker != "" {
+			uploads = trimAfterUploadIDAndKey(p.KeyMarker, p.UploadIDMarker, uploads)
+		} else {
+			uploads = trimAfterUploadKey(p.KeyMarker, uploads)
+		}
+	}
+
+	if len(uploads) > p.MaxUploads {
+		result.IsTruncated = true
+		uploads = uploads[:p.MaxUploads]
+		result.NextUploadIDMarker = uploads[len(uploads)-1].UploadID
+		result.NextKeyMarker = uploads[len(uploads)-1].Key
+	}
+
+	uniqDirs := make(map[string]struct{})
+
+	for _, ov := range uploads {
+		if ov.IsDir {
+			if _, ok := uniqDirs[ov.Key]; !ok {
+				uniqDirs[ov.Key] = struct{}{}
+				result.Prefixes = append(result.Prefixes, ov.Key)
+			}
```

Comment:
I suppose we should check directory uniqueness before truncate up to `MaxUploads`